File: boj4949.java
import java.util.*; 
import java.io.*; 
public class boj4949{ // change to public class Main for submission. 
	public static void main(String[] args){
		Scanner sc = new Scanner(System.in);  
		String strLine; // read line by line
		while ((strLine = sc.nextLine()) != null){
			if (strLine.equals(".")) break; 
			Stack<Character> stack = new Stack<>();    
			boolean cannot = false; 
			for (int i = 0; i < strLine.length(); i++){
				if (strLine.charAt(i) == '(' || strLine.charAt(i) == '['){
					stack.push(strLine.charAt(i)); 
				}else if (strLine.charAt(i) == ')'){
					if (stack.isEmpty()){
						cannot = true;  
						System.out.println("no"); 
						break; 
					}else if (stack.peek() == '('){
						stack.pop(); 
					}else{
						cannot = true; 
						System.out.println("no"); 
						break; 
					}
				}else if (strLine.charAt(i) == ']'){
					if (stack.isEmpty()){
						cannot = true; 
						System.out.println("no"); 
						break; 
					}else if (stack.peek() == '['){
						stack.pop(); 
					}else{
						cannot = true; 
						System.out.println("no"); 
						break; 
					}
				}
			}
			if (!cannot){
				if (stack.isEmpty()) System.out.println("yes");  
				else System.out.println("no"); 
			}  
		}
	}
}

File: boj4949.cpp
#include <iostream>
#include <cstdlib>
#include <stack>
#include <algorithm>
using namespace std;  
int main(int argc,char **argv){
	string s;  
	while (getline(cin,s)){
		if (s == ".") break; 
		stack<char> st;  
		bool cannot = false;  
		for (int i = 0; i < s.size(); i++){
			if (s[i] == '(' || s[i] == '['){
				st.push(s[i]);  
			}else if (s[i] == ')'){
				if (st.empty()){
					cannot = true; 
					cout << "no" << endl; 
					break;  
				}else if (st.top() == '('){
					st.pop();  
				}else{
					cannot = true;  
					cout << "no" << endl; 
					break; 
				}
			}else if (s[i] == ']'){
				if (st.empty()){
					cannot = true; 
					cout << "no" << endl; 
					break; 
				}else if (st.top() == '['){
					st.pop();  
				}else{
					cannot = true;  
					cout << "no" << endl; 
					break; 
				}
			}
		}
		if (!cannot){
			cout << (st.empty() ? "yes" : "no") << endl; 
		}
	}
	return 0; 
}
